portfolio/src/main/java/com/google/sps/servlets/CommentsController.java

```diff
@@ -14,36 +14,239 @@
+        upperIndex <= count ? upperIndex : count);
+
+    Comments comments = new Comments(info, trimmedComments);
+    Gson gson = new Gson();
+    String data = gson.toJson(comments);
+
+    response.getWriter().println(data);
   }
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("application/json;");
 
+    String body = getBody(request);
+    JSONObject json = new JSONObject();
+    try {
+      JSONParser parser = new JSONParser();
+      json = (JSONObject) parser.parse(body);
+    } catch (ParseException e) {
+      e.printStackTrace();
+    }
+
+    String username = getAttribute(json, "username", "Anonymous");
+    String comment = getAttribute(json, "comment", "...");
+    long createdAt = System.currentTimeMillis();
+    int likes = 0;
+
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("username", username);
+    commentEntity.setProperty("comment", comment);
+    commentEntity.setProperty("createdAt", createdAt);
+    commentEntity.setProperty("likes", likes);
+
+    datastore.put(commentEntity);
+
+    long id = commentEntity.getKey().getId();
+
     Gson gson = new Gson();
-    String data = gson.toJson(comments);
+    String data = gson.toJson(new Comment(id, username, comment, createdAt, likes));
+
+    response.getWriter().println(data);
+  }
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json;");
+
+    String body = getBody(request);
+    JSONObject json = new JSONObject();
+
+    try {
+      JSONParser parser = new JSONParser();
+      json = (JSONObject) parser.parse(body);
+    } catch (ParseException e) {
+      e.printStackTrace();
+    }
+
+    String id = getAttribute(json, "id", null);
+    Key key = KeyFactory.createKey("Comment", Long.parseLong(id));
+
+    Entity commentEntity;
+    try {
+      commentEntity = datastore.get(key);
+    } catch (EntityNotFoundException e) {
+      e.printStackTrace();
+      return;
+    }
+
+    int likes = ((Long) commentEntity.getProperty("likes")).intValue();
+    likes++;
+    commentEntity.setProperty("likes", likes);
+
+    datastore.put(commentEntity);
+
+    String username = (String) commentEntity.getProperty("username");
+    String comment = (String) commentEntity.getProperty("comment");
+    long createdAt = (long) commentEntity.getProperty("createdAt");
 
-    response.getWriter().println(data);  
+    Gson gson = new Gson();
+    String data = gson.toJson(new Comment(commentEntity.getKey().getId(), username, comment, createdAt, likes));
+
+    response.getWriter().println(data);
+  }
+  
+  @Override
+  public void doDelete(HttpServletRequest request, HttpServletResponse response) throws IOException {
```

Comment:
Although the tutorial recommended setting the delete functionality in another servlet, I don't see why the `DELETE` method wasn't mentioned, so I decided to set that in this same controller.

Reply:
Sounds good!